hail/src/main/scala/is/hail/expr/types/physical/PArray.scala

```diff
@@ -39,4 +39,42 @@ final case class PArray(elementType: PType, override val required: Boolean = fal
     assert(this isOfType other)
     CodeOrdering.iterableOrdering(this, other.asInstanceOf[PArray], mb)
   }
+
+  def checkedConvertFrom(mb: EmitMethodBuilder, r: Code[Region], value: Code[_], otherPT: PType, msg: String): Code[Long] = {
```

Comment:
I might be wrong, but it looks like this method could be moved up to `PContainer`. It might also be nice to factor out a method
```scala
PContainer.initWithElements(aoff: Code[Long], length: Code[Int], elementsOff: Code[Long])
```
Then the non-trivial branch here could simplify to
```scala
val newOffset = mb.newField[Long]
Code(
  if (otherPTA.elementType.required)
    Code.empty()
  else {
    val i = mb.newField[Int]
    i := 0,
    Code.whileLoop(i < len,
      otherPTA.isElementMissing(oldOffset, i).orEmpty(Code._fatal(s"${msg}: convertFrom $otherPT failed: element missing.")),
      i := i + 1),
  },
  newOffset := allocate(r, len),
  initWithElements(newOffset, len, otherPTA.elementOffset(oldOffset, len, 0)),
  newOffset)
```